### DISCUSSION TITLE
Start points without end points in pickup&delivery problem

### DISCUSSION THREAD

**Discussion by CatonTheRoRo** [google/or-tools#4405] [Routing (and legacy CP) questions]
Is there a way to specify start points without end points, So the solver would not take endpoints as constrain in his solving process?
When I simply omit endpoints I got 'wrong number of arguments' error.
Is passing endpoints to RoutingIndexManager mandatory?
Is there a way or way-around to make or-tools run without endpoints?

**Comment by jmarca** [comment]
Can you give a code snippet example of how you are calling this?

Guessing...

If *all* of your pickups do not have delivery points, then you do not have a pickup and delivery problem.

If just *some* of your pickups do not have delivery points, then those should be *excluded* from all of the code you use to specify the pdp pairs.

```py
for item in items:
    if item.delivery == null:
        continue
    do_pdp_stuff(item)
```

Another idea...if all of the pickup without delivery items are supposed to be taken back to the depot, then you can create dummy delivery nodes for each item at the depot.

Somethig like, in magical pseudo python:

```python
# create needed dummy nodes prior to setting up the routing object
for item in items:
    if item.delivery == null:
        # make a dummy node
        dummy_nodes += 1
        item.delivery = dummy_nodes
        lookup_thing[dummy_nodes] = depot_node

... later ...

for item in items:
    do_pdp_stuff(item)

```

James

On Wed, Oct 09, 2024 at 01:47:39PM -0700, CatonTheRoRo wrote:
> Is there a way to specify start points without end points, So the solver would not take endpoints as constrain in his solving process?
> When I simply omit endpoints I got 'wrong number of arguments' error.
> Is passing endpoints to RoutingIndexManager mandatory?
> Is there a way or way-around to make or-tools run without endpoints?
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/4405
> You are receiving this because you are subscribed to this thread.
> 
> Message ID: ***@***.***>

-- 

James E. Marca
Activimetrics LLC

**Reply by CatonTheRoRo** [reply]
I use the example in or-tools page: https://github.com/google/or-tools/blob/stable/ortools/constraint_solver/samples/vrp_pickup_delivery.py
There is no depot, by 'start point', I mean the current driver start point:
```python
data["pickups_deliveries"] = [
        [3, 4],
        [2, 5]
    ]
    data["num_vehicles"] = 2
    data["starts"] = [0, 1]
    data["ends"] = [0, 1]
```
The index manager function call:
```python
manager = pywrapcp.RoutingIndexManager(
        len(data["distance_matrix"]), data["num_vehicles"], data["starts"], data["ends"]
    )
```
All pickups have deliveries, I was pointing to start and end points of drivers locations.
If I remove endpoints argument from the index manager:
```python
manager = pywrapcp.RoutingIndexManager(
        len(data["distance_matrix"]), data["num_vehicles"], data["starts"]
    )
```
I get wrong number of arguments error.

**Comment by lperron** [comment]
The question of no end points has been answered multiple times. 
Have you tried the suggestions ?

**Reply by CatonTheRoRo** [reply]
I am posting this after researching multiple times here and in other places. I did not find an answer that meets my needs. If you find one, please send me the link.

**Comment by lperron** [comment]
as indicated in this thread, you did not search hard
https://groups.google.com/g/or-tools-discuss/c/tSM-f5GzWds

There is even a guide entry for this:

https://developers.google.com/optimization/routing/routing_tasks#allowing_arbitrary_start_and_end_locations


Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le jeu. 10 oct. 2024 à 14:14, CatonTheRoRo ***@***.***> a
écrit :

**Reply by CatonTheRoRo** [reply]
I had already found those before. I read them again and then spotted this phrase: '(and modify the solution printer to omit the depot).' OMG—my problem was already solved, but I got confused by the printed output. Anyway, thank you for your help.

**Comment by Pratibha-UID** [comment]
Could you please help me out where I am doing mistake, as I have followed the suggestion where I want to specify the start point and end should decided by algorithm for that I have tried same way which you mentioned in Allowing arbitrary start and end locations documentation but in result I am getting random start point and start point which I really expect at start that came at last location.

RoutingIndexManager(distancemetrix.length, numVehicles, new int[]{1}, new int[]{0});
numVehicles = 1

Below distance metrix I have passed.
Before :
{              {0, 972105, 889576, 968248, 987336, 883578, 934372},
                {971997, 0, 95095, 5255, 17380, 99897, 68607},
                {890190, 94882, 0, 92999, 108008, 6093, 59123},
                {969866, 5273, 92964, 0, 17566, 97766, 66476},
                {986885, 17967, 109983, 17203, 0, 114785, 83495},
                {882710, 100256, 6953, 98734, 113743, 0, 64858},
                {937247, 68597, 60345, 66714, 81723, 65147, 0}}
After adding arbitrary end point  :
{               {0,0,0,0,0,0,0},
                {0, 972105, 889576, 968248, 987336, 883578, 934372},
                {0, 0, 95095, 5255, 17380, 99897, 68607},
                {0, 94882, 0, 92999, 108008, 6093, 59123},
                {0, 5273, 92964, 0, 17566, 97766, 66476},
                {0, 17967, 109983, 17203, 0, 114785, 83495},
                {0, 100256, 6953, 98734, 113743, 0, 64858},
                {0, 68597, 60345, 66714, 81723, 65147, 0}}
                
Can anyone help on this where I am doing wrong?
